### PR TITLE
Feature/moderator privilege

### PR DESCRIPTION
**Pull request by Vishtar**

### CODE REVIEW

File: services/servers/utils/getLatestMemes.js [hunk 1]
```diff
@@ -3,15 +3,28 @@ import {
...
 
-export const getLatestMemes = async (client, from, to, size) => {
+export const getLatestMemes = async (client, from, to, size, filtersString) => {
+    const filterObject = filtersString ? JSON.parse(filtersString) : null;
+    const additionalFilter = {};
+    if (filterObject?.channel?.length) {
+        ((additionalFilter.must ??= {}).terms ??= {}).channelName = filterObject.channel.map(channel => channel.toLowerCase());
+    }
+    if (filterObject?.not?.state !== null) {
+        ((additionalFilter.must_not ??= {}).term ??= {}).state = filterObject?.state ?? 1;
```
Comment (klntsky) [inline services/servers/utils/getLatestMemes.js:13]: >        ((additionalFilter.must_not ??= {})

You've just set the variable. You know that `additionaFilter.must_not` is `undefined`. Don't check for it, it obscures the intention

Comment (klntsky) [reply services/servers/utils/getLatestMemes.js:13]: Same 3 lines above

File: services/servers/utils/getLatestMemes.js [hunk 2]
```diff
@@ -3,15 +3,28 @@ import {
 }  from '../../../constants/index.js';
 import { getMemeResponseEntity } from './index.js';
 
-export const getLatestMemes = async (client, from, to, size) => {
+export const getLatestMemes = async (client, from, to, size, filtersString) => {
```
Comment (klntsky) [inline services/servers/utils/getLatestMemes.js:6]: `filtersString` does not follow a particular schema. It's bad to pass arbitrary filters to elastic from the frontend due to the possibility of DOS.

File: services/servers/utils/searchMemes.js [hunk 3]
```diff
@@ -13,6 +13,13 @@ export const searchMemes = async (client, query, page, size) => {
         size,
         query: {
             bool: {
+                must_not: [
+                    {
+                        match: {
+                            state: 1, // hidden
```
Comment (klntsky) [inline services/servers/utils/searchMemes.js:19]:
```suggestion
                            state: 1, // we hide soft-deleted memes
```

File: services/servers/web/routers/admin/methods/channelMemesStatePut.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { setChannelMemesState } from '../../../../utils/index.js';
+import { setLogAction } from '../utils/index.js';
+
+export const channelMemesStatePut = async (req, res) => {
+    const client = req.app.get('elasticClient');
+    const { channel, state } = req.body;
+    if (!channel || !state)
+        return res.status(500).send();
+    await setChannelMemesState(client, channel, state);
```
Comment (klntsky) [inline services/servers/web/routers/admin/methods/channelMemesStatePut.js:9]: Again, you are passing arbitrary json to elastic.

File: services/servers/web/routers/admin/methods/channelMemesStatePut.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { setChannelMemesState } from '../../../../utils/index.js';
+import { setLogAction } from '../utils/index.js';
+
+export const channelMemesStatePut = async (req, res) => {
```
Comment (klntsky) [inline services/servers/web/routers/admin/methods/channelMemesStatePut.js:4]: Where's the authentication?

File: services/servers/web/index.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    const response = await getLatestMemes(client, from, to, SEARCH_PAGE_SIZE, filters);
+    return res.send(response);
+});
+
+app.get('/getChannelList', async (req, res) => {
+    const { page, onlyAvailable, filter } = req.query;
+    const mysql = await getMysqlClient();
+    const filters = [];
+    if (onlyAvailable === 'true') filters.push('availability IS TRUE');
+    if (filter) filters.push(`name LIKE "%${filter}%"`);
```
Comment (klntsky) [inline services/servers/web/index.js:98]: Do not build mysql queries here. Mysql should not leak into the domain logic.

Comment (klntsky) [inline services/servers/web/index.js:98]: Why not ILIKE?

File: utils/mysql-queries/getChannels.js [hunk 7]
```diff
@@ -1,9 +1,14 @@
-export const getChannels = async (mysql, page, onlyAvailable, size) => {
-    const filter = onlyAvailable ? 'WHERE availability IS TRUE' : '';
+export const getChannels = async (mysql, page, size, filters) => {
     const offset = (page - 1) * size;
     const [results] = await mysql.query(`
         SELECT name, availability FROM channels
-        ${filter}
+        ${filters?.length
+            ? `WHERE ${filters.join(' AND ')}`
```
Comment (klntsky) [inline utils/mysql-queries/getChannels.js:6]: Do not use string interpolation in SQL.

File: utils/mysql-queries/getChannelsCount.js [hunk 8]
```diff
@@ -1,8 +1,10 @@
-export const getChannelsCount = async (mysql, onlyAvailable) => {
-    const filter = onlyAvailable ? 'WHERE availability IS TRUE' : '';
+export const getChannelsCount = async (mysql, filters) => {
     const [[results]] = await mysql.query(`
         SELECT COUNT(*) FROM channels
-        ${filter}
+        ${filters?.length
+            ? `WHERE ${filters.join(' AND ')}`
```
Comment (klntsky) [inline utils/mysql-queries/getChannelsCount.js:5]: Do not use string interpolation in SQL.

**Review by klntsky — CHANGES_REQUESTED** [review]

File: services/servers/web/routers/admin/methods/channelMemesStatePut.js (hunk 5 above)
Comment (Vishtar) [reply services/servers/web/routers/admin/methods/channelMemesStatePut.js:4]: It's on top of the entire adminRouter (/admin/* methods).

File: utils/mysql-queries/getChannels.js (hunk 7 above)
Comment (Vishtar) [reply utils/mysql-queries/getChannels.js]: I wrapped this in a [mysql.escapeId()](https://github.com/mysqljs/mysql?tab=readme-ov-file#escaping-query-identifiers).

File: services/servers/utils/getLatestMemes.js (hunk 2 above)
Comment (Vishtar) [reply services/servers/utils/getLatestMemes.js:6]: I'm not sure that I understand the problem right. Now this is checking for channel and state variable types.

**Pull request merged**